y3prediction/prediction.py

```diff
@@ -2467,11 +2819,181 @@ def unfiltered_rhs(t, state):
             ignition_rhs = ignition_source(x_vec=x_vec, state=fluid_state,
                                            eos=gas_model.eos, time=t)/current_dt
 
+        av_smu_rhs = 0*cv.mass
+        av_sbeta_rhs = 0*cv.mass
+        av_skappa_rhs = 0*cv.mass
+        av_smu_wall = 0*wv.mass
+        av_sbeta_wall = 0*wv.mass
+        av_skappa_wall = 0*wv.mass
+        av_smu_wall_rhs = 0*wv.mass
+        av_sbeta_wall_rhs = 0*wv.mass
+        av_skappa_wall_rhs = 0*wv.mass
+        #eta = 1.e-10
+        # work good for shock 1d
+        tau = 1.e-6
+        eta = 0.1*tau
+        epsilon_diff = eta/tau
+        #epsilon_diff = 0.
+
+        if use_av > 0:
+            # regular boundaries for smoothness mu
+            smooth_neumann = NeumannDiffusionBoundary(0)
+            fluid_av_smu_boundaries = {
+                flow_bnd.domain_tag: smooth_neumann,
+                wall_bnd.domain_tag: smooth_neumann,
+            }
+            wall_av_smu_boundaries = {
+                wall_ffld_bnd.domain_tag: smooth_neumann,
+            }
+
+            # interface boundaries for smoothness mu
+            pairwise_diff_smu = {
+                (dd_vol_fluid, dd_vol_wall):
+                    (av_smu, av_smu_wall)}
+            pairwise_diff_smu_tpairs = inter_volume_trace_pairs(
+                dcoll, pairwise_diff_smu, comm_tag=_MuDiffCommTag)
+            diff_smu_tpairs = pairwise_diff_smu_tpairs[dd_vol_fluid, dd_vol_wall]
+
+            wall_av_smu_boundaries.update({
+                tpair.dd.domain_tag:
+                NeumannDiffusionBoundary(0)
+                for tpair in diff_smu_tpairs})
+
+            reverse_diff_smu_tpairs = pairwise_diff_smu_tpairs[dd_vol_wall,
+                                                               dd_vol_fluid]
+
+            fluid_av_smu_boundaries.update({
+                tpair.dd.domain_tag:
+                NeumannDiffusionBoundary(0)
+                for tpair in reverse_diff_smu_tpairs})
+
+            # av mu
+            av_smu_rhs = (
+                diffusion_operator(
+                    dcoll, epsilon_diff, fluid_av_smu_boundaries, av_smu,
+                    quadrature_tag=quadrature_tag, dd=dd_vol_fluid,
+                    comm_tag=_MuDiffFluidCommTag
+                ) + 1/tau * (smoothness_mu - av_smu)
+            )
+
+            # reverse in the wall for matching
+            av_smu_wall_rhs = (
+                diffusion_operator(
+                    dcoll, 0., wall_av_smu_boundaries, av_smu_wall,
+                    quadrature_tag=quadrature_tag, dd=dd_vol_wall,
+                    comm_tag=_MuDiffWallCommTag
+                )
+            )
```

Comment:
I added a `filter_part_boundaries` function to `grudge.discretization` that might help simplify this. With it, you should be able to do something like:

```python
            from grudge.discretization import filter_part_boundaries
            fluid_av_smu_boundaries.update({
                 dd_bdry.domain_tag: NeumannDiffusionBoundary(0)
                 for dd_bdry in filter_part_boundaries(
                     dcoll, volume_dd=dd_vol_fluid, neighbor_volume_dd=dd_vol_wall)})

             # av mu
             av_smu_rhs = (
                 diffusion_operator(
                     dcoll, epsilon_diff, fluid_av_smu_boundaries, av_smu,
                     quadrature_tag=quadrature_tag, dd=dd_vol_fluid,
                     comm_tag=_MuDiffFluidCommTag
                 ) + 1/tau * (smoothness_mu - av_smu)
             )
```
And then you don't need to do the reverse stuff, because you're not exchanging data across the interface anymore. (This doesn't apply to the characteristic length, since you're still using external data there.)